src/cow-react/modules/limitOrders/pure/DeadlineSelector/index.tsx

```diff
@@ -0,0 +1,94 @@
+import { Dropdown } from '@cow/common/pure/Dropdown'
+import { LimitOrderDeadline, limitOrdersDeadlines, maxCustomDeadline } from './deadlines'
+
+import { useCallback, useMemo, useRef } from 'react'
+import { ChevronDown } from 'react-feather'
+import * as styledEl from './styled'
+import ms from 'ms.macro'
+import { Trans } from '@lingui/macro'
+
+function limitDateString(date: Date): string {
+  const [first, second] = date.toISOString().split(':')
+
+  return [first, second].join(':')
+}
+
+const customDateOptions: Intl.DateTimeFormatOptions = {
+  year: '2-digit',
+  month: 'short',
+  day: 'numeric',
+  hour: '2-digit',
+  minute: '2-digit',
+  hour12: false,
+}
+
+export interface DeadlineSelectorProps {
+  deadline: LimitOrderDeadline | undefined
+  customDeadline: number | null
+  selectDeadline(deadline: LimitOrderDeadline): void
+  selectCustomDeadline(deadline: number): void
+}
+
+export function DeadlineSelector(props: DeadlineSelectorProps) {
+  const { deadline, customDeadline, selectDeadline, selectCustomDeadline } = props
+  const currentDeadlineNode = useRef<HTMLButtonElement>()
+
+  const min = limitDateString(new Date(Date.now() + ms`10m`))
+  const max = limitDateString(new Date(Date.now() + maxCustomDeadline))
+
+  const existingDeadline = useMemo(() => {
+    return limitOrdersDeadlines.find((item) => item === deadline)
+  }, [deadline])
+
+  const customDeadlineTitle = useMemo(() => {
+    if (!customDeadline) return ''
+    return new Date(customDeadline).toLocaleString(undefined, customDateOptions)
+  }, [customDeadline])
+
+  const setDeadline = useCallback(
+    (deadline: LimitOrderDeadline) => {
+      selectDeadline(deadline)
+      currentDeadlineNode.current?.click() // Close dropdown
+    },
+    [selectDeadline]
+  )
+
+  const onChange = useCallback(
+    (event) => {
+      const customDeadline = new Date(event.target.value).getTime()
+
+      selectCustomDeadline(customDeadline)
+    },
+    [selectCustomDeadline]
+  )
+
+  const list = (
+    <styledEl.ListWrapper>
+      {limitOrdersDeadlines.map((item) => (
+        <li key={item.value}>
+          <styledEl.ListItem onClick={() => setDeadline(item)}>
+            <Trans>{item.title}</Trans>
+          </styledEl.ListItem>
+        </li>
+      ))}
+      <styledEl.ListItem>
+        <Trans>Custom</Trans>
+        <styledEl.CustomInput type="datetime-local" onChange={onChange} min={min} max={max} />
+      </styledEl.ListItem>
+    </styledEl.ListWrapper>
+  )
+
+  return (
+    <styledEl.Wrapper>
+      <styledEl.Title>
+        <Trans>Expiry</Trans>
+      </styledEl.Title>
+      <Dropdown content={list}>
+        <styledEl.Current ref={currentDeadlineNode as any} isCustom={!!customDeadline}>
```

Comment:
just curious, do we really need the `any` here?

Reply:
Thanks!
Casted to `RefObject<HTMLButtonElement>`